File: src/benchmark.c
#include "benchmark.h"
#include "archetypes.h"
#include "ecs_entity.h"
#include "logger.h"

#include <stdlib.h>

// FPS floor we ramp up to.
#define BENCH_MIN_FPS 30.0
// Steady-state entity count to settle at after the peak.
#define BENCH_TARGET 1
// How long to average FPS before making a spawn/despawn decision. Long enough
// to smooth out per-frame jitter, short enough to feel responsive.
#define BENCH_SAMPLE_SEC 0.4
// Entities added/removed per second, spread evenly across frames. Spreading the
// work avoids a big single-frame spawn burst that would otherwise spike the
// frame time and make the benchmark measure spawn cost instead of sim cost.
// Kept moderate so the count climbs gradually and is easy to watch.
#define BENCH_SPAWN_RATE 1500.0
// Pause at the steady state before automatically restarting the run (seconds).
#define BENCH_RESTART_DELAY 1.5

Benchmark* benchmark_new(void)
{
    Benchmark *b = calloc(1, sizeof(Benchmark));
    return b;
}

void benchmark_free(Benchmark *b)
{
    if (!b) return;
    free(b->ids);
    free(b);
}

static void bench_push(Benchmark *b, Entity e)
{
    if (b->count == b->capacity) {
        size_t nc = b->capacity ? b->capacity * 2 : 64;
        Entity *p = realloc(b->ids, nc * sizeof(Entity));
        if (!p) return;
        b->ids = p;
        b->capacity = nc;
    }
    b->ids[b->count++] = e;
}

static void bench_spawn(struct Game *G, size_t n)
{
    for (size_t i = 0; i < n; i++) {
        Entity e = ECS_EntityNew(G->ecs, G->archetypes[TEST_BOUNCING_SPRITE_ARCHETYPE]);
        bench_push(G->bench, e);
    }
}

static void bench_despawn(struct Game *G, size_t n)
{
    Benchmark *b = G->bench;
    for (size_t i = 0; i < n && b->count > 0; i++) {
        Entity e = b->ids[--b->count];
        ECS_EntityDelete(G->ecs, e);
    }
}

void benchmark_start(struct Game *G)
{
    Benchmark *b = G->bench;

    bench_despawn(G, b->count);
    b->phase = BENCH_RAMP_UP;
    b->peak = 0;
    b->prev_count = 0;
    b->accum_time = 0;
    b->accum_frames = 0;
    b->smoothed_fps = 0;
    b->done_time = 0;
    b->active = true;

    bench_spawn(G, 1); // start with a single object
    LOG_INFO("Benchmark started (ramp up to %.0f fps floor)", BENCH_MIN_FPS);
}

void benchmark_stop(struct Game *G)
{
    Benchmark *b = G->bench;
    if (!b->active) return;

    bench_despawn(G, b->count);
    b->active = false;
}

void benchmark_update(struct Game *G)
{
    Benchmark *b = G->bench;
    if (!b->active) return;

    // Finished: hold at the steady state briefly, then loop the whole run.
    if (b->phase == BENCH_DONE) {
        b->done_time += (double)G->dtime;
        if (b->done_time >= BENCH_RESTART_DELAY) {
            LOG_INFO("Benchmark restarting...");
            benchmark_start(G);
        }
        return;
    }

    // Spread spawning/despawning evenly across frames at a constant rate, so a
    // single frame never does a big burst. This keeps the per-frame cost flat
    // and makes the FPS reflect steady-state simulation, not spawn cost.
    size_t rate = (size_t)(BENCH_SPAWN_RATE * (double)G->dtime) + 1;

    if (b->phase == BENCH_RAMP_UP) {
        bench_spawn(G, rate);
    } else if (b->phase == BENCH_RAMP_DOWN && b->count > BENCH_TARGET) {
        size_t togo = b->count - BENCH_TARGET;
        bench_despawn(G, rate < togo ? rate : togo);
    }

    // Build up a sampling window for a stable FPS reading, then act on it.
    b->accum_time += G->dtime;
    b->accum_frames++;
    if (b->accum_time < BENCH_SAMPLE_SEC) return;

    b->smoothed_fps = b->accum_frames / b->accum_time;
    b->accum_time = 0;
    b->accum_frames = 0;

    switch (b->phase) {
    case BENCH_RAMP_UP:
        if (b->smoothed_fps <= BENCH_MIN_FPS) {
            // prev_count is the count at the previous sample, which still held
            // the floor; report it as the peak (avoids within-window overshoot).
            b->peak = b->prev_count ? b->prev_count : b->count;
            b->phase = BENCH_RAMP_DOWN;
            LOG_INFO("Benchmark peak: %zu entities sustained >= %.0f fps",
                     b->peak, BENCH_MIN_FPS);
        } else {
            b->prev_count = b->count;
        }
        break;

    case BENCH_RAMP_DOWN:
        if (b->count <= BENCH_TARGET) {
            b->phase = BENCH_DONE;
            LOG_INFO("Benchmark done. Steady state: %zu entities (~%.0f fps)",
                     b->count, b->smoothed_fps);
        }
        break;

    case BENCH_DONE:
    default:
        break;
    }
}

File: scripts/benchmark.lua
-- scripts/benchmark.lua
--
-- Adaptive ECS stress benchmark (ported from C to Lua). It ramps the entity
-- count up until the frame rate hits a floor (recording the peak), ramps back
-- down to a single object, pauses, then loops.
--
-- Only the *policy* (how many to spawn/despawn this frame) lives here. The heavy
-- per-frame work -- moving and bouncing every sprite -- stays in the C
-- VelocitySystem, so the measured throughput reflects the engine, not Lua.

local MIN_FPS      = 30.0     -- frame-rate floor we ramp up to
local TARGET       = 1        -- settle count after the peak
local SAMPLE_SEC   = 0.4      -- FPS averaging window
local SPAWN_RATE   = 1500.0   -- entities added/removed per second (frame-spread)
local RESTART_WAIT = 1.5      -- pause at the floor before looping

-- No Collision component: these sprites are pure load and must not feed the
-- collision system. Empty Transform/Velocity = the C constructors' random
-- position and velocity.
prefab "Bench" {
  Transform = {},
  Velocity  = {},
  Sprite    = { image = "skeleton" },
}

local phase     -- "up" | "down" | "done"
local n         -- entities alive
local peak
local prev_n
local accum_t, accum_f
local done_t

local function reset()
  if n and n > 0 then despawn("Bench", n) end
  phase, n, peak, prev_n = "up", 0, 0, 0
  accum_t, accum_f, done_t = 0, 0, 0
  spawn_many("Bench", 1)   -- start with a single object
  n = 1
end

function on_start()
  reset()
end

function on_update(dt)
  if phase == "done" then
    done_t = done_t + dt
    if done_t >= RESTART_WAIT then reset() end
    return
  end

  -- Per-frame path is kept allocation-free (no string building): spread
  -- spawning/despawning evenly across frames so no single frame bursts (which
  -- would measure spawn cost instead of steady-state simulation).
  local rate = math.floor(SPAWN_RATE * dt) + 1
  if phase == "up" then
    spawn_many("Bench", rate)
    n = n + rate
  elseif phase == "down" and n > TARGET then
    n = n - despawn("Bench", math.min(rate, n - TARGET))
  end

  -- FPS sampling window for a stable reading. The HUD string is only built here
  -- (~2-3 times/sec), never per frame, so it doesn't churn the GC under load.
  accum_t = accum_t + dt
  accum_f = accum_f + 1
  if accum_t < SAMPLE_SEC then return end
  local smoothed = accum_f / accum_t
  accum_t, accum_f = 0, 0

  if phase == "up" then
    hud("BENCHMARK: ramping up... " .. n)
    if smoothed <= MIN_FPS then
      -- prev_n held the floor at the previous sample; report it as the peak
      -- (avoids within-window overshoot).
      peak  = (prev_n > 0) and prev_n or n
      phase = "down"
      log(string.format("Benchmark peak: %d entities >= %d fps", peak, MIN_FPS))
    else
      prev_n = n
    end
  elseif phase == "down" then
    hud(string.format("BENCHMARK: peak %d @%dfps, settling (%d)", peak, MIN_FPS, n))
    if n <= TARGET then
      phase = "done"
      hud(string.format("BENCHMARK: peak %d entities @%dfps", peak, MIN_FPS))
      log(string.format("Benchmark done. Settled at %d (~%.0f fps)", n, smoothed))
    end
  end
end
